### PR TITLE
fn: size restricted tmpfs /tmp and read-only / support

### PR DESCRIPTION
**Pull request by skinowski**
*) read-only Root Fs Support
*) removed CPUShares from docker API. This was unused.
*) docker.Prepare() refactoring
*) added docker.configureTmpFs() for size limited tmpfs on /tmp
*) tmpfs size support in routes and resource tracker
*) fix fn-test-utils to handle sparse files better in create file

### CODE REVIEW

**Comment by zootalures** [comment]
Is it feasible to to add an e2e test to witness the FS limit here?

**Comment by zootalures** [comment]
I have some minor concerns about the impact of making this per-route configurable  - but understand that while this is necessarily tmpfs that this make sense.

**Comment by rdallman** [comment]
seems like a good use case for annotations to configure ?

File: api/agent/call.go [hunk 1]
```diff
@@ -252,7 +253,9 @@ func (a *agent) GetCall(opts ...CallOpt) (Call, error) {
 		return nil, errors.New("no model or request provided for call")
 	}
 
-	if !a.resources.IsResourcePossible(c.Memory, uint64(c.CPUs), c.Type == models.TypeAsync) {
+	mem := c.Memory + uint64(c.TmpFsSize)
```
Comment (kmjohansen) [inline api/agent/call.go]: I was originally on the fence about whether we should account for this, but after looking at `runc` and the kernel's `memcg` implementation, I think this is mandatory for correct behavior.  The cgroup's total page allocation count is charged against the memory limit in the memory cgroup, so if we under provision here then the user won't be able to use all of tmpfs.  The werid thing is that this lets them consume more of any pages, so if they have a piggy app they could use all of their allocation plus our extra for tmp, and then not get everything available for transient storage.  That said, file backed and regular anon pages are also charged against this total memory cap, so the behavior at the limit is going to seem strange no matter what.

File: api/agent/config.go [hunk 2]
```diff
@@ -102,6 +107,10 @@ func NewAgentConfig() (*AgentConfig, error) {
 		cfg.EnableNBResourceTracker = true
 	}
 
+	if _, ok := os.LookupEnv(EnvEnableReadOnlyRootFs); ok {
```
Comment (kmjohansen) [inline api/agent/config.go]: Does this imply that we need a change in the pure runner to enable this feature?

**Comment by skinowski** [comment]
Annotations sounds good. Then operators could choose to put this as app or route based setting.

File: api/agent/config.go (hunk 2 above)
Comment (skinowski) [reply api/agent/config.go]: Yes, runner env would need an update.

File: api/agent/call.go (hunk 1 above)
Comment (skinowski) [reply api/agent/call.go]: Yeah, I felt this is a bit aggressive here as well, but safer than over committing IMHO.

**Comment by skinowski** [comment]
@rdallman FYI, @zootalures has some reservations on using annotations for this. Leaving this PR as is, I think per route config is OK and gives enough flexibility to operators.

**Comment by kmjohansen** [comment]
Are there any corresponding changes that need to be made to the fn-api or cli to support this?  Relatedly: should we enable by default and instead use the environment to allow users to disable?

**Comment by skinowski** [comment]
I'd say we leave it disabled by default. This is the least disruptive for open-source users, cli, etc.

**Pull request merged**